onnxruntime/core/framework/bfc_arena.h

```diff
@@ -476,6 +485,29 @@ class BFCArena : public IAllocator {
 
   ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(BFCArena);
 };
+
+class StreamAwareArena : public BFCArena {
+public:
+  StreamAwareArena(std::unique_ptr<IAllocator> resource_allocator,
+                   size_t total_memory,
+                   bool enable_dynamic_cross_stream_sharing,
+                   ArenaExtendStrategy arena_extend_strategy = DEFAULT_ARENA_EXTEND_STRATEGY,
+                   int initial_chunk_size_bytes = DEFAULT_INITIAL_CHUNK_SIZE_BYTES,
+                   int max_dead_bytes_per_chunk = DEFAULT_MAX_DEAD_BYTES_PER_CHUNK,
+                   int initial_growth_chunk_size_bytes = DEFAULT_INITIAL_GROWTH_CHUNK_SIZE_BYTES);
+
+  //If size is 0, then this function returns either NULL,
+  //or a unique pointer value that can later be successfully
+  //passed to free(). Whatever, do not dereference that pointer
+  void* AllocOnStream(size_t size, Stream* current_stream_id, WaitNotificationFn wait_fn);
+
+  void ReleaseStreamBuffers(Stream* stream);
+
+  StreamAwareArena* AsStreamAwareAreana() override;
```

Comment:
Not sure why I would call AsStreamAwareAreana on a StreamAwareArena instance?

Reply:
The design is the ArenaAllocator has a public API "AsStreamAwareAreana", for the arena that does not support stream, it will return nullptr, otherwise it return this. I think it is more efficient than a dynamic_cast. does it make sense?